compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion/MetadataEncoding.hs

```diff
@@ -82,6 +93,40 @@ mapFunDep f (a, b) = (map f a, map f b)
 mapFunDepM :: Monad m => (a -> m b) -> (GHC.FunDep a -> m (GHC.FunDep b))
 mapFunDepM f (a, b) = liftM2 (,) (mapM f a) (mapM f b)
 
+---------------------
+-- MINIMAL PRAGMAS --
+---------------------
+
+minimalName :: LF.TypeSynName -> LF.ExprValName
+minimalName (LF.TypeSynName xs) = LF.ExprValName ("$$minimal" <> T.concat xs)
+
+pattern TEncodedStr :: T.Text -> LF.Type
+pattern TEncodedStr x = LF.TStruct [(LF.FieldName x, LF.TUnit)]
+
+pattern TEncodedCon :: T.Text -> LF.Type -> LF.Type
+pattern TEncodedCon a b = LF.TStruct [(LF.FieldName a, b)]
+
+encodeLBooleanFormula :: BF.LBooleanFormula T.Text -> LF.Type
+encodeLBooleanFormula = encodeBooleanFormula . GHC.unLoc
+
+decodeLBooleanFormula :: LF.Type -> Maybe (BF.LBooleanFormula T.Text)
+decodeLBooleanFormula = fmap GHC.noLoc . decodeBooleanFormula
+
+encodeBooleanFormula :: BF.BooleanFormula T.Text -> LF.Type
```

Comment:
The MINIMAL pragma can be loaded with arbitrary positive Boolean formulae? Wow, I didn't know that. It's also a bit scary. What do you think about "flattening" the formula into disjunctive normal form and encode the result as a list of lists of strings? I would assume the blowup coming with conversion to DNF doesn't matter here in practice and it would allow us to drop the `And`, `Or`, and `Parens` thingies.

Reply:
Yeah. I think I prefer to leave the boolean formulas in place. GHC already does some simplifying, but it doesn't go full DNF for good reasons, I assume (e.g. better error messages). Also I don't want to risk introducing new errors by normalizing the boolean formula incorrectly.